File: packages/db_models/command_centre.py
"""Peacock Command Centre — flagship visibility command snapshot.

Surfaces Peacock Visibility Index dimensions, situation briefing, and an
intelligence feed of PEACOCK DETECTED events — not a generic SEO dashboard.
"""

from __future__ import annotations

from datetime import datetime

from sqlalchemy import (
    DateTime,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from db_models.base import Base, WorkspaceTenantMixin


VISIBILITY_DIMENSIONS: tuple[str, ...] = (
    "search_visibility",
    "ai_visibility",
    "share_of_answer",
    "entity_authority",
    "citation_authority",
    "content_opportunity",
    "agent_readiness",
)

VISIBILITY_LABELS: dict[str, str] = {
    "search_visibility": "Search Visibility",
    "ai_visibility": "AI Visibility",
    "share_of_answer": "Share of Answer",
    "entity_authority": "Entity Authority",
    "citation_authority": "Citation Authority",
    "content_opportunity": "Content Opportunity",
    "agent_readiness": "Agent Readiness",
}

SITUATION_KINDS: tuple[str, ...] = (
    "biggest_opportunity",
    "biggest_threat",
    "fastest_win",
    "competitor_movement",
    "ai_visibility_change",
    "critical_technical_issue",
)

SITUATION_LABELS: dict[str, str] = {
    "biggest_opportunity": "Biggest Opportunity",
    "biggest_threat": "Biggest Threat",
    "fastest_win": "Fastest Win",
    "competitor_movement": "Competitor Movement",
    "ai_visibility_change": "AI Visibility Change",
    "critical_technical_issue": "Critical Technical Issue",
}

METHODOLOGY = "peacock_command_centre_v1"
METHODOLOGY_NOTE = (
    "Peacock Command Centre is the flagship command surface for generative "
    "visibility intelligence. It centres the Peacock Visibility Index, a "
    "situation briefing, and a PEACOCK DETECTED intelligence feed — not a "
    "generic SEO metric dashboard."
)


class CommandCentreSnapshot(Base, WorkspaceTenantMixin):
    """One Command Centre snapshot for a brand/website."""

    __tablename__ = "command_centre_snapshots"

    website_id: Mapped[str] = mapped_column(
        ForeignKey("websites.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    client_brand: Mapped[str] = mapped_column(String(255), nullable=False, index=True)
    snapshot_status: Mapped[str] = mapped_column(
        String(32), default="completed", nullable=False, index=True
    )
    methodology: Mapped[str] = mapped_column(String(64), default=METHODOLOGY, nullable=False)
    visibility_index: Mapped[float] = mapped_column(Float, nullable=False)
    visibility_delta: Mapped[float] = mapped_column(Float, nullable=False)
    captured_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    headline: Mapped[str] = mapped_column(String(255), nullable=False)
    summary: Mapped[str | None] = mapped_column(Text)
    notes: Mapped[str | None] = mapped_column(Text)

    signals: Mapped[list[CcVisibilitySignal]] = relationship(
        back_populates="snapshot", cascade="all, delete-orphan", passive_deletes=True
    )
    situations: Mapped[list[CcSituationItem]] = relationship(
        back_populates="snapshot", cascade="all, delete-orphan", passive_deletes=True
    )
    feed_items: Mapped[list[CcFeedItem]] = relationship(
        back_populates="snapshot", cascade="all, delete-orphan", passive_deletes=True
    )


class CcVisibilitySignal(Base, WorkspaceTenantMixin):
    """One Peacock Visibility Index dimension."""

    __tablename__ = "cc_visibility_signals"
    __table_args__ = (UniqueConstraint("snapshot_id", "dimension"),)

    snapshot_id: Mapped[str] = mapped_column(
        ForeignKey("command_centre_snapshots.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    dimension: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    label: Mapped[str] = mapped_column(String(128), nullable=False)
    score: Mapped[float] = mapped_column(Float, nullable=False)
    delta: Mapped[float] = mapped_column(Float, nullable=False)
    rank_order: Mapped[int] = mapped_column(Integer, nullable=False)

    snapshot: Mapped[CommandCentreSnapshot] = relationship(back_populates="signals")


class CcSituationItem(Base, WorkspaceTenantMixin):
    """Second-layer situation briefing item."""

    __tablename__ = "cc_situation_items"
    __table_args__ = (UniqueConstraint("snapshot_id", "kind"),)

    snapshot_id: Mapped[str] = mapped_column(
        ForeignKey("command_centre_snapshots.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    kind: Mapped[str] = mapped_column(String(64), nullable=False, index=True)
    label: Mapped[str] = mapped_column(String(128), nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    detail: Mapped[str] = mapped_column(Text, nullable=False)
    severity: Mapped[str] = mapped_column(String(16), nullable=False, index=True)
    rank_order: Mapped[int] = mapped_column(Integer, nullable=False)

    snapshot: Mapped[CommandCentreSnapshot] = relationship(back_populates="situations")


class CcFeedItem(Base, WorkspaceTenantMixin):
    """Intelligence feed detection (PEACOCK DETECTED)."""

    __tablename__ = "cc_feed_items"
    __table_args__ = (UniqueConstraint("snapshot_id", "feed_index"),)

    snapshot_id: Mapped[str] = mapped_column(
        ForeignKey("command_centre_snapshots.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    feed_index: Mapped[int] = mapped_column(Integer, nullable=False)
    detection_label: Mapped[str] = mapped_column(
        String(64), default="PEACOCK DETECTED", nullable=False
    )
    headline: Mapped[str] = mapped_column(String(255), nullable=False)
    body: Mapped[str] = mapped_column(Text, nullable=False)
    primary_driver: Mapped[str] = mapped_column(Text, nullable=False)
    potential_response: Mapped[str] = mapped_column(Text, nullable=False)
    confidence: Mapped[float] = mapped_column(Float, nullable=False)
    detected_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), nullable=False, index=True
    )
    graph_surface: Mapped[str | None] = mapped_column(String(64), index=True)

    snapshot: Mapped[CommandCentreSnapshot] = relationship(back_populates="feed_items")

File: Peacock/frontend/src/lib/command-centre.ts
import { getApiBaseUrl } from "@/lib/api";

export type VisibilitySignal = {
  dimension: string;
  label: string;
  score: number;
  delta: number;
  rank_order: number;
};

export type SituationItem = {
  kind: string;
  label: string;
  title: string;
  detail: string;
  severity: string;
  rank_order: number;
};

export type FeedItem = {
  feed_index: number;
  detection_label: string;
  headline: string;
  body: string;
  primary_driver: string;
  potential_response: string;
  confidence: number;
  confidence_pct: number;
  detected_at: string;
  graph_surface: string | null;
};

export type CommandCentreSnapshot = {
  client_brand: string;
  visibility_index: number;
  visibility_delta: number;
  captured_at: string;
  headline: string;
  signals: VisibilitySignal[];
  situations: SituationItem[];
  feed_items: FeedItem[];
  methodology_note: string;
  summary: string;
};

export const DEMO_SNAPSHOT: CommandCentreSnapshot = {
  client_brand: "Acme",
  visibility_index: 59.4,
  visibility_delta: -0.6,
  captured_at: new Date().toISOString(),
  headline: "Acme · Peacock Visibility Index 59",
  methodology_note:
    "Peacock Command Centre is the flagship command surface for generative visibility intelligence.",
  summary: "Command Centre demo snapshot.",
  signals: [
    { dimension: "search_visibility", label: "Search Visibility", score: 72, delta: 1.4, rank_order: 0 },
    { dimension: "ai_visibility", label: "AI Visibility", score: 58, delta: -3.8, rank_order: 1 },
    { dimension: "share_of_answer", label: "Share of Answer", score: 41, delta: -2.1, rank_order: 2 },
    { dimension: "entity_authority", label: "Entity Authority", score: 63, delta: 0.6, rank_order: 3 },
    { dimension: "citation_authority", label: "Citation Authority", score: 47, delta: -5.2, rank_order: 4 },
    { dimension: "content_opportunity", label: "Content Opportunity", score: 81, delta: 4.0, rank_order: 5 },
    { dimension: "agent_readiness", label: "Agent Readiness", score: 54, delta: 1.1, rank_order: 6 },
  ],
  situations: [
    {
      kind: "biggest_opportunity",
      label: "Biggest Opportunity",
      title: "Proprietary benchmark study",
      detail:
        "Publishing an owned benchmark can reclaim citation share Acme lost on commercial prompts.",
      severity: "high",
      rank_order: 0,
    },
    {
      kind: "biggest_threat",
      label: "Biggest Threat",
      title: "Competitor A citation surge",
      detail: "Competitor A citation share jumped 18% → 31% on category research queries.",
      severity: "critical",
      rank_order: 1,
    },
    {
      kind: "fastest_win",
      label: "Fastest Win",
      title: "Refresh /compare + /pricing hubs",
      detail: "Entity-dense comparison tables are the shortest path to SoA recovery this sprint.",
      severity: "high",
      rank_order: 2,
    },
    {
      kind: "competitor_movement",
      label: "Competitor Movement",
      title: "Competitor A shipped 3 research pages",
      detail: "Those pages are the primary driver behind the citation-share acceleration.",
      severity: "high",
      rank_order: 3,
    },
    {
      kind: "ai_visibility_change",
      label: "AI Visibility Change",
      title: "AI Visibility −3.8 this week",
      detail: "Claude and Perplexity presence softened; ChatGPT held flatter.",
      severity: "medium",
      rank_order: 4,
    },
    {
      kind: "critical_technical_issue",
      label: "Critical Technical Issue",
      title: "Indexation gap on /guides/*",
      detail:
        "Agent and crawler readiness checks flag thin schema + blocked snippets on guide templates.",
      severity: "critical",
      rank_order: 5,
    },
  ],
  feed_items: [
    {
      feed_index: 0,
      detection_label: "PEACOCK DETECTED",
      headline: "Competitor A increased citation share",
      body: "Competitor A increased citation share from 18% → 31%.",
      primary_driver: "3 recently published research pages.",
      potential_response: "Publish proprietary benchmark study.",
      confidence: 0.87,
      confidence_pct: 87,
      detected_at: new Date().toISOString(),
      graph_surface: "citation_graph",
    },
    {
      feed_index: 1,
      detection_label: "PEACOCK DETECTED",
      headline: "AI Visibility dipped across two engines",
      body: "Share of Answer softened on Claude (−4pp) and Perplexity (−2pp) week-over-week.",
      primary_driver: "Citation disappearance on commercial prompt cluster.",
      potential_response: "Reinforce comparison hubs with quotable specs and sources.",
      confidence: 0.81,
      confidence_pct: 81,
      detected_at: new Date().toISOString(),
      graph_surface: "anomaly_engine",
    },
    {
      feed_index: 2,
      detection_label: "PEACOCK DETECTED",
      headline: "Content opportunity cluster opened on /security",
      body: "Opportunity Engine ranks /security in the top GEO improvement set.",
      primary_driver: "Entity Authority gap on ‘enterprise reliability’ facet.",
      potential_response: "Assign evidence-dense writer; ship FAQ + source block variant.",
      confidence: 0.76,
      confidence_pct: 76,
      detected_at: new Date().toISOString(),
      graph_surface: "opportunity_engine",
    },
  ],
};

export async function fetchCommandCentrePreview(
  brand = "Acme",
): Promise<CommandCentreSnapshot> {
  try {
    const response = await fetch(
      `${getApiBaseUrl()}/command-centre/preview?brand=${encodeURIComponent(brand)}`,
      { cache: "no-store" },
    );
    if (!response.ok) {
      return DEMO_SNAPSHOT;
    }
    return (await response.json()) as CommandCentreSnapshot;
  } catch {
    return DEMO_SNAPSHOT;
  }
}
